File: mockserver-client-dotnet/test/MockServer.Client.Tests/CallbackTests.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using FluentAssertions;
using MockServer.Client.Models;
using Xunit;

namespace MockServer.Client.Tests;

/// <summary>
/// Tests for class callbacks (REST serialization) and object/closure callbacks
/// (WebSocket reply-envelope routing via direct ProcessMessage, mirroring BreakpointTests).
/// </summary>
public class CallbackTests
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    // --- Class callback serialization ---

    [Fact]
    public void HttpResponseClassCallback_Serializes_Correctly()
    {
        var expectation = new Expectation
        {
            HttpRequest = HttpRequest.Request().WithMethod("GET").WithPath("/api").Build(),
            HttpResponseClassCallback = new HttpClassCallback
            {
                CallbackClass = "com.example.MyCallback",
                Delay = new Delay { TimeUnit = TimeUnit.MILLISECONDS, Value = 50 },
                Primary = true
            }
        };

        var json = JsonSerializer.Serialize(expectation, JsonOptions);
        var parsed = JsonNode.Parse(json)!;

        parsed["httpResponseClassCallback"]!["callbackClass"]!.GetValue<string>()
            .Should().Be("com.example.MyCallback");
        parsed["httpResponseClassCallback"]!["delay"]!["value"]!.GetValue<long>().Should().Be(50);
        parsed["httpResponseClassCallback"]!["primary"]!.GetValue<bool>().Should().BeTrue();
    }

    [Fact]
    public void HttpForwardClassCallback_Serializes_Correctly()
    {
        var expectation = new Expectation
        {
            HttpRequest = HttpRequest.Request().WithPath("/fwd").Build(),
            HttpForwardClassCallback = HttpClassCallback.Of("com.example.FwdCallback")
        };

        var json = JsonSerializer.Serialize(expectation, JsonOptions);
        var parsed = JsonNode.Parse(json)!;

        parsed["httpForwardClassCallback"]!["callbackClass"]!.GetValue<string>()
            .Should().Be("com.example.FwdCallback");
        // optional fields omitted when null
        parsed["httpForwardClassCallback"]!.AsObject().ContainsKey("delay").Should().BeFalse();
        parsed["httpForwardClassCallback"]!.AsObject().ContainsKey("primary").Should().BeFalse();
    }

    [Fact]
    public void ClassCallback_OmitsNull_DelayAndPrimary()
    {
        var json = JsonSerializer.Serialize(HttpClassCallback.Of("c.X"), JsonOptions);
        json.Should().Contain("\"callbackClass\":\"c.X\"");
        json.Should().NotContain("delay");
        json.Should().NotContain("primary");
    }

    // --- Object callback serialization ---

    [Fact]
    public void HttpResponseObjectCallback_Serializes_ClientId()
    {
        var expectation = new Expectation
        {
            HttpRequest = HttpRequest.Request().WithPath("/obj").Build(),
            HttpResponseObjectCallback = new HttpObjectCallback { ClientId = "ws-client-1" }
        };

        var json = JsonSerializer.Serialize(expectation, JsonOptions);
        var parsed = JsonNode.Parse(json)!;

        parsed["httpResponseObjectCallback"]!["clientId"]!.GetValue<string>().Should().Be("ws-client-1");
        // responseCallback/delay/primary omitted when null
        parsed["httpResponseObjectCallback"]!.AsObject().ContainsKey("responseCallback").Should().BeFalse();
    }

    // --- Object callback WS routing: reply envelope ---

    [Fact]
    public void ObjectResponseCallback_Produces_HttpResponse_Reply_EchoingCorrelationId()
    {
        var ws = new BreakpointWebSocketClient();
        JsonObject? capturedReply = null;
        string? capturedType = null;

        ws.SetSendInterceptorForTest((type, valueJson) =>
        {
            capturedType = type;
            capturedReply = JsonNode.Parse(valueJson)!.AsObject();
        });

        ws.SetObjectResponseHandler(request =>
        {
            var path = request["path"]?.GetValue<string>();
            return new JsonObject
            {
                ["statusCode"] = 200,
                ["body"] = $"handled {path}"
            };
        });

        // A request frame WITHOUT X-MockServer-BreakpointId is an object-callback dispatch.
        var request = new JsonObject
        {
            ["method"] = "GET",
            ["path"] = "/dynamic",
            ["headers"] = new JsonObject
            {
                ["WebSocketCorrelationId"] = new JsonArray("corr-obj-1")
            }
        };
        var envelope = new JsonObject
        {
            ["type"] = "org.mockserver.model.HttpRequest",
            ["value"] = request.ToJsonString()
        };

        ws.ProcessMessage(envelope.ToJsonString());

        capturedType.Should().Be("org.mockserver.model.HttpResponse");
        capturedReply.Should().NotBeNull();
        capturedReply!["statusCode"]!.GetValue<int>().Should().Be(200);
        capturedReply["body"]!.GetValue<string>().Should().Be("handled /dynamic");
        // WebSocketCorrelationId echoed back as a header array
        capturedReply["headers"]!["WebSocketCorrelationId"]!.AsArray()[0]!.GetValue<string>()
            .Should().Be("corr-obj-1");
    }

    [Fact]
    public void ObjectForwardCallback_Produces_HttpRequest_Reply()
    {
        var ws = new BreakpointWebSocketClient();
        string? capturedType = null;
        JsonObject? capturedReply = null;

        ws.SetSendInterceptorForTest((type, valueJson) =>
        {
            capturedType = type;
            capturedReply = JsonNode.Parse(valueJson)!.AsObject();
        });

        ws.SetObjectForwardHandler(request =>
        {
            request["path"] = "/rewritten";
            return request;
        });

        var request = new JsonObject
        {
            ["method"] = "GET",
            ["path"] = "/original",
            ["headers"] = new JsonObject
            {
                ["WebSocketCorrelationId"] = new JsonArray("corr-fwd-1")
            }
        };
        var envelope = new JsonObject
        {
            ["type"] = "org.mockserver.model.HttpRequest",
            ["value"] = request.ToJsonString()
        };

        ws.ProcessMessage(envelope.ToJsonString());

        capturedType.Should().Be("org.mockserver.model.HttpRequest");
        capturedReply!["path"]!.GetValue<string>().Should().Be("/rewritten");
        capturedReply["headers"]!["WebSocketCorrelationId"]!.AsArray()[0]!.GetValue<string>()
            .Should().Be("corr-fwd-1");
    }

    [Fact]
    public void BreakpointRequest_Still_Routes_To_BreakpointHandler_NotObjectCallback()
    {
        var ws = new BreakpointWebSocketClient();
        bool breakpointHandled = false;
        bool objectHandled = false;

        ws.SetRequestHandler("bp-1", req => { breakpointHandled = true; return req; });
        ws.SetObjectResponseHandler(req => { objectHandled = true; return new JsonObject { ["statusCode"] = 200 }; });

        // Frame WITH a breakpoint id must go to the breakpoint handler.
        var request = new JsonObject
        {
            ["path"] = "/bp",
            ["headers"] = new JsonObject
            {
                ["WebSocketCorrelationId"] = new JsonArray("corr-bp"),
                ["X-MockServer-BreakpointId"] = new JsonArray("bp-1")
            }
        };
        ws.ProcessMessage(new JsonObject
        {
            ["type"] = "org.mockserver.model.HttpRequest",
            ["value"] = request.ToJsonString()
        }.ToJsonString());

        breakpointHandled.Should().BeTrue();
        objectHandled.Should().BeFalse();
    }
}

File: mockserver-client-rust/tests/callback_tests.rs
//! Unit tests for callback models and object-callback message routing.
//! These run with NO server — they test serialization of class/object callbacks
//! on the Expectation model and the object-callback reply-envelope routing.

use mockserver_client::*;
use std::sync::{Arc, Mutex};

// ---------------------------------------------------------------------------
// HttpClassCallback serialization
// ---------------------------------------------------------------------------

#[test]
fn test_class_callback_serializes_callback_class() {
    let cb = HttpClassCallback::new("com.example.MyCallback");
    let json = serde_json::to_value(&cb).unwrap();
    assert_eq!(json["callbackClass"], "com.example.MyCallback");
    // Optional fields omitted when unset.
    assert!(json.get("delay").is_none());
    assert!(json.get("primary").is_none());
}

#[test]
fn test_class_callback_serializes_delay_and_primary() {
    let cb = HttpClassCallback::new("com.example.MyCallback")
        .delay(Delay::milliseconds(250))
        .primary(true);
    let json = serde_json::to_value(&cb).unwrap();
    assert_eq!(json["callbackClass"], "com.example.MyCallback");
    assert_eq!(json["delay"]["timeUnit"], "MILLISECONDS");
    assert_eq!(json["delay"]["value"], 250);
    assert_eq!(json["primary"], true);
}

#[test]
fn test_expectation_response_class_callback_field_name() {
    let exp = Expectation::new(HttpRequest::new().method("GET").path("/cb"))
        .respond_with_class_callback("com.example.ResponseCallback");
    let json = serde_json::to_value(&exp).unwrap();
    assert_eq!(
        json["httpResponseClassCallback"]["callbackClass"],
        "com.example.ResponseCallback"
    );
    // Must use the camelCase wire field name, not the snake_case Rust field.
    assert!(json.get("http_response_class_callback").is_none());
}

#[test]
fn test_expectation_forward_class_callback_field_name() {
    let exp = Expectation::new(HttpRequest::new().method("GET").path("/fwd"))
        .forward_with_class_callback("com.example.ForwardCallback");
    let json = serde_json::to_value(&exp).unwrap();
    assert_eq!(
        json["httpForwardClassCallback"]["callbackClass"],
        "com.example.ForwardCallback"
    );
}

#[test]
fn test_class_callback_builder_from_value() {
    let exp = Expectation::new(HttpRequest::new().path("/x")).respond_class_callback(
        HttpClassCallback::new("com.example.C").primary(false),
    );
    let json = serde_json::to_value(&exp).unwrap();
    assert_eq!(json["httpResponseClassCallback"]["callbackClass"], "com.example.C");
    assert_eq!(json["httpResponseClassCallback"]["primary"], false);
}

// ---------------------------------------------------------------------------
// HttpObjectCallback serialization
// ---------------------------------------------------------------------------

#[test]
fn test_object_callback_serializes_client_id() {
    let cb = HttpObjectCallback::new("client-abc");
    let json = serde_json::to_value(&cb).unwrap();
    assert_eq!(json["clientId"], "client-abc");
    assert!(json.get("delay").is_none());
}

#[test]
fn test_expectation_response_object_callback_field_name() {
    let exp = Expectation::new(HttpRequest::new().method("GET").path("/obj"))
        .respond_object_callback(HttpObjectCallback::new("client-xyz"));
    let json = serde_json::to_value(&exp).unwrap();
    assert_eq!(json["httpResponseObjectCallback"]["clientId"], "client-xyz");
    assert!(json.get("http_response_object_callback").is_none());
}

#[test]
fn test_expectation_forward_object_callback_field_name() {
    let exp = Expectation::new(HttpRequest::new().path("/obj-fwd"))
        .forward_object_callback(HttpObjectCallback::new("fwd-client"));
    let json = serde_json::to_value(&exp).unwrap();
    assert_eq!(json["httpForwardObjectCallback"]["clientId"], "fwd-client");
}

// ---------------------------------------------------------------------------
// Object-callback routing (no live WebSocket needed)
// ---------------------------------------------------------------------------

fn make_object_request_json(correlation_id: &str) -> String {
    serde_json::json!({
        "method": "GET",
        "path": "/echo",
        "headers": {
            "WebSocketCorrelationId": [correlation_id]
        }
    })
    .to_string()
}

fn make_breakpoint_request_json(breakpoint_id: &str, correlation_id: &str) -> String {
    serde_json::json!({
        "method": "GET",
        "path": "/echo",
        "headers": {
            "WebSocketCorrelationId": [correlation_id],
            "X-MockServer-BreakpointId": [breakpoint_id]
        }
    })
    .to_string()
}

#[test]
fn test_route_object_callback_produces_response_envelope() {
    let handler: Arc<Mutex<Option<ObjectResponseHandler>>> = Arc::new(Mutex::new(Some(Box::new(
        |req: serde_json::Value| {
            let path = req["path"].as_str().unwrap_or("");
            serde_json::json!({
                "statusCode": 200,
                "body": format!("echo {path}")
            })
        },
    ))));

    let json = make_object_request_json("corr-obj-1");
    let result = route_object_callback(&json, &handler);
    assert!(result.is_some());
    let (type_name, value) = result.unwrap();

    // Reply must be an HttpResponse envelope type.
    assert_eq!(type_name, "org.mockserver.model.HttpResponse");
    // The closure-derived response is returned ...
    assert_eq!(value["statusCode"], 200);
    assert_eq!(value["body"], "echo /echo");
    // ... and the WebSocketCorrelationId is echoed back so the server can route it.
    assert_eq!(
        extract_header(&value, "WebSocketCorrelationId"),
        Some("corr-obj-1".to_string())
    );
}

#[test]
fn test_route_object_callback_ignores_breakpoint_frames() {
    // Even with a handler registered, a frame carrying a breakpoint id must NOT
    // be object-routed — it belongs to the breakpoint path.
    let handler: Arc<Mutex<Option<ObjectResponseHandler>>> = Arc::new(Mutex::new(Some(Box::new(
        |_req: serde_json::Value| serde_json::json!({ "statusCode": 200 }),
    ))));

    let json = make_breakpoint_request_json("bp-1", "corr-bp");
    assert!(route_object_callback(&json, &handler).is_none());
}

#[test]
fn test_route_object_callback_none_when_no_handler() {
    let handler: Arc<Mutex<Option<ObjectResponseHandler>>> = Arc::new(Mutex::new(None));
    let json = make_object_request_json("corr-none");
    assert!(route_object_callback(&json, &handler).is_none());
}

#[test]
fn test_route_object_callback_panic_auto_continues() {
    let handler: Arc<Mutex<Option<ObjectResponseHandler>>> = Arc::new(Mutex::new(Some(Box::new(
        |_req: serde_json::Value| panic!("boom"),
    ))));

    let json = make_object_request_json("corr-panic");
    let result = route_object_callback(&json, &handler);
    // On panic we still produce a (passthrough) response envelope with the
    // correlation id echoed, so the exchange is not left hanging.
    assert!(result.is_some());
    let (type_name, value) = result.unwrap();
    assert_eq!(type_name, "org.mockserver.model.HttpResponse");
    assert_eq!(
        extract_header(&value, "WebSocketCorrelationId"),
        Some("corr-panic".to_string())
    );
}
